bootloader/piboot.md

```diff
@@ -0,0 +1,127 @@
+   kernel will be a matter of extracting/creating these assets and
+   setting `os_prefix` appropriately.
+1. On kernel refresh, we will create a `tryboot.txt` file pointing to
+   the new assets in its `os_prefix` setting, and reboot with a "0
+   tryboot" parameter so we use `tryboot.txt` after rebooting. The
+   `cmdline.txt` in the `os_prefix` folder will contain a special
+   parameter (`kernel_status=trying`) so UC knows that we are using
+   the configuration from `tryboot.txt`. With this, the OS can
+   understand if the new kernel has been used to boot.
+
+To keep state we have environment files that, differently to grub or
+U-Boot, are not directly read or written by the bootloader, but are
+instead translated to bootloader configuration files when kernel
+updates happen. These environment files are named `piboot.conf` and
+live in the `piboot/ubuntu/` folder inside ubuntu-seed partition. In
+run mode, the folder from the boot partition would be mounted in the
+`/boot/piboot/` directory. This is analogous to what is done by UC for
+grub and U-Boot environment files.
+
+### Snapd
+
+The piboot environment files have the same format as U-Boot
+environment files so we leverage the existing snapd codebase and take
+advantage of error detection capabilities. We store there key/value
+paris. The `GetBootVars()`/`SetBootVars()` methods for the bootloader
+interface read and modify these files in the usual way. However, these
+files do not directly affect the bootloader any more, so we need to
+generate the bootloader configuration files when some of the variables
+are changed, that is, when `SetBootVars()` is called it will re-create
+the bootloader configuration depending on what has changed. The method
+that implements the changes will have as input the environment file in
+the seed partition, and will generate `{config,tryboot}.txt` and
+`cmdline.txt` accordingly.
+
+The logic that changes the `kernel_status` variable in other
+bootloaders (see `bootloader/assets/data/grub.cfg`) cannot be done by
+piboot as it does not support scripting and is closed source. Instead,
+basically the same thing has been implemented in the initramfs, and it
+is run as part of the snap bootstrap code (see
+`boot.updatePibootKernelStatus()` function). Running it from the
+initramfs ensures that the code is run only once in boot. This code
+looks at the kernel command and checks if `kernel_status=trying` is
+present to change the status in the environment file to `try`. Once
+snapd starts, it will check and set `kernel_status` in the usual way
+for any bootloader.
+
+### RPi gadget
+
+Gadgets of the “bootloader: piboot” type need to have
+
+1. An empty `piboot.conf file` (which contains environment variables
+   as explained above, and will be filled with values by snapd during
+   image preparation, installation and kernel updates)
+1. Reference configuration file for the bootloader, named
+   `config.templ.txt`. It will be used to create `config.txt`, with a
```

Comment:
do we still require the config.templ.txt file?

Reply:
Nope, good catch